cmd/deploy.go

```diff
@@ -15,107 +15,395 @@
+	serviceName := projectName
+	deployImage := projectName // if not tagged, this is the deploy image name
+	if tag != "" {
+		deployImage = tag //Otherwise, it's the tag
+	}
+	// We're not pushing to a repository, so we need to use dev.local for Knative to be able to find it
+	if !push {
+		localtag := "dev.local/" + projectName
+		// Tagging the image using the tag as the deployImage for KNative
+		err = DockerTag(deployImage, localtag)
+		if err != nil {
+			return errors.Errorf("Tagging the image failed - exiting. Error: %v", err)
+		}
+		deployImage = localtag // And forcing deployimage to be localtag
+	}
+	//Generating the KNative yaml file
+	Debug.logf("Calling GenKnativeYaml with parms: %s %d %s %s \n", knativeTempl, port, serviceName, deployImage)
+	yamlFileName, err := GenKnativeYaml(knativeTempl, port, serviceName, deployImage, push)
+	if err != nil {
+		return errors.Errorf("Could not generate the KNative YAML file: %v", err)
+	}
+	Info.log("Generated KNative serving deploy file: ", yamlFileName)
+	// Pushing the docker image if necessary
+	if push {
+		err = DockerPush(deployImage)
+		if err != nil {
+			return errors.Errorf("Could not push the docker image - exiting. Error: %v", err)
+		}
+	}
+	err = KubeApply(yamlFileName)
+	// Performing the kubectl apply
+	if err != nil {
+		return errors.Errorf("Failed to deploy to your Kubernetes cluster: %v", err)
+	}
+	Info.log("Deployment succeeded.")
+	url, err := KubeGetKnativeURL(serviceName)
+	if err != nil {
+		return errors.Errorf("Failed to find deployed service in your Kubernetes cluster: %v", err)
+	}
+	Info.log("Your deployed service is available at the following URL: ", url)
+
+	return nil
+}
+
+func generateDeploymentConfig() error {
+	containerConfigDir := "/config/app-deploy.yaml"
+
+	exists, err := exists(configFile)
+	if err != nil {
+		return errors.Errorf("Error checking status of %s", configFile)
+	}
+	if exists && !force {
+		return errors.Errorf("Error, deploy config file %s already exists. Specify an alternative file using --file or using --force to overwrite.", configFile)
+	}
+
+	setupErr := setupConfig()
+	if setupErr != nil {
+		return setupErr
+	}
+	projectConfig, configErr := getProjectConfig()
+	if configErr != nil {
+		return configErr
+	}
+	err = CheckPrereqs()
+	if err != nil {
+		Warning.logf("Failed to check prerequisites: %v\n", err)
+	}
+
+	stackImage := projectConfig.Platform
+	Debug.log("Stack image: ", stackImage)
+	Debug.log("Config directory: ", containerConfigDir)
+
+	var cmdName string
+	var cmdArgs []string
+	dockerPullErr := dockerPullImage(stackImage)
+	if dockerPullErr != nil {
+		return dockerPullErr
+	}
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		err := dockerStop(containerName)
+		if err != nil {
+			Error.log(err)
+		}
+		//dockerRemove(containerName) is not needed due to --rm flag
+		os.Exit(1)
+	}()
+	cmdName = "docker"
```

Comment:
Future: no change needed, this code looks like the code we have in extract, I wonder if there is a potential here for future refactoring?